sqlmigrations: persist table desc format upgrades

benesch opened the pull request:
See commit messages for details. The last commit is the one that really matters:

```
Add a migration that ensures that the upgrade to InterleavedFormatVersion is
persisted to disk for all table descriptors. It must otherwise be performed
on-the-fly whenever a table descriptor is loaded. In fact, before this
migration, a cluster that was continuously upgraded from before beta-20161013
would retain its old-format table descriptors until a schema-mutating statement
was executed against every old-format table.
```

The earlier commits are mostly yak-shaving.

@danhhz you probably want to look at just the last commit.

benesch committed:
sqlmigrations: test correctness of ExpectedInitialRanges

This computation is finicky and easy to get wrong. If the computation
underestimates the expected number of ranges the error is nearly
impossible to detect, as WaitForInitialSplits simply won't wait quite
long enough.

Add a test to verify its computation is correct.

Release note: None

benesch committed:
sqlmigrations,server: simplify computation of ExpectedInitialRangeCount

Computing how many splits a migration will cause is highly dependent on
ordering; a migration that adds a new system table with ID 19 will
create only one split if the prior system table ID is 18, but five
splits if the prior system table ID is 14, for example, because we split
at every possible table ID even if no table exists with that ID. This
makes it very difficult to remove/reorder migrations in tests and
requires a bunch of explanatory comments when tables skip IDs.

Fix this up by teaching each migration to record the IDs of the
descriptors that it will add instead of the number of ranges/descriptors
it will create. Then we can centralize computation of exactly how this
will impact the number of splits in ExpectedInitialRanges.

Release note: None

benesch committed:
sqlmigrations: extract framework for isolated migration testing

Pay down some technical debt by wrapping this error-prone package-global
mutation pattern into a reusable test helper.

Release note: None

benesch committed:
sqlmigrations: persist table desc format upgrades

Add a migration that ensures that the upgrade to
InterleavedFormatVersion is persisted to disk for all table descriptors.
It must otherwise be performed on-the-fly whenever a table descriptor is
loaded. In fact, before this migration, a cluster that was continuously
upgraded from before beta-20161013 would retain its old-format table
descriptors until a schema-mutating statement was executed against every
old-format table.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20701)
<!-- Reviewable:end -->

danhhz commented:
lgtm but :man_scientist: :dog:

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sqlmigrations/migrations.go, line 563 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20701#-L0GtEfbcpqs2hjmOFUF:-L0GtEfbcpqs2hjmOFUG:b-hzvbln) ([raw file](https://github.com/cockroachdb/cockroach/blob/237b860aeed0c8b9909237aca3b0ec5793533385/pkg/sqlmigrations/migrations.go#L563)):*
> ```Go
> 				if table := sqlDesc.GetTable(); table != nil {
> 					if upgraded := table.MaybeUpgradeFormatVersion(); upgraded {
> 						table.UpVersion = true
> ```

tableDesc.SetUpVersion errors if the table is being dropped, so we should probably handle that case here (I assume by skipping that TableDescriptor?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20701#-:-L0GtWsuBjJmh5u184mL:b-a37v4d)*
<!-- Sent from Reviewable.io -->

benesch commented:
Addressed your comment below. @a-robinson, will definitely want your eyes on the new `IsolatedMigrationTest` framework I added.

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sqlmigrations/migrations.go, line 563 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20701#-L0GtEfbcpqs2hjmOFUF:-L0ROnCkPXOX9DGaPnBy:b-ipmho1) ([raw file](https://github.com/cockroachdb/cockroach/blob/237b860aeed0c8b9909237aca3b0ec5793533385/pkg/sqlmigrations/migrations.go#L563)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

tableDesc.SetUpVersion errors if the table is being dropped, so we should probably handle that case here (I assume by skipping that TableDescriptor?)
</blockquote></details>

We actually _need_ to make sure to upgrade the dropped table descriptors too. I added a nice long explanation about why, plus a test to verify that it's actually safe.

(The error in `SetUpVersion` seems to exist to prevent accidental `ALTER TABLE`s on dropped table descriptors, not because setting the flag on a dropped table is actually unsafe.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20701#-:-L0RQ0d5jMiYIJBSwVvR:b-i44qdp)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Sorry for the delay!

---

Reviewed 1 of 2 files at r1, 5 of 5 files at r5, 5 of 5 files at r6, 1 of 1 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/drop_test.go, line 674 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20701#-L0gT2HEQ8ms_lU7OBXZ:-L0gT2HEQ8ms_lU7OBX_:b-u6v29s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4e398e3a3ad79ab3eb7a1ae81db671f8b92b93/pkg/sql/drop_test.go#L674)):*
> ```Go
> 
> 	s, sqlDBRaw, kvDB := serverutils.StartServer(t, params)
> 	defer s.Stopper().Stop(context.TODO())
> ```

Just use the `ctx` already defined above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20701#-:-L0gUQKvH9BQ8eLmyJ__:b-8yidu2)*
<!-- Sent from Reviewable.io -->

benesch commented:
No problem, @a-robinson. TFTR!

mberhault commented:
Nice refactor! That should make migrations easier to deal with.

benesch commented:
Why thanks! :) I think I'll have a fix for some of this backfill pain soon too.
